securedrop/management/sources.py

```diff
@@ -0,0 +1,58 @@
+import argparse
+from typing import List
+
+from db import db
+from encryption import EncryptionManager, GpgKeyNotFoundError
+from management import app_context
+from models import Source
+
+
+def remove_pending_sources(args: argparse.Namespace) -> int:
+    """
+    Removes pending source accounts, with the option of keeping
+    the n newest source accounts.
+    """
+    n = args.keep_most_recent
+    sources = find_pending_sources(n)
+    print(f"Found {len(sources)} pending sources")
+
+    deleted = []
+    for source in sources:
+        try:
+            EncryptionManager.get_default().delete_source_key_pair(source.filesystem_id)
+        except GpgKeyNotFoundError:
+            pass
+        delete_pending_source(source)
+        deleted.append(source)
+
+    print(f"Deleted {len(sources)} pending sources")
+    return 0
+
+
+def find_pending_sources(keep_most_recent: int) -> List[Source]:
+    """
+    Finds all sources that are marked as pending
+    """
+    with app_context():
+        pending_sources = (
+            Source.query.filter_by(pending=True)
+            .order_by(Source.id.desc())
+            .offset(keep_most_recent)
+            .all()
+        )
+
+    return pending_sources
+
+
+def delete_pending_source(source: Source) -> None:
+    """
+    Delete a pending source from the database
+    """
+    if source.pending:
+        with app_context():
+            try:
+                db.session.delete(source)
+                db.session.commit()
+            except Exception as exc:
```

Comment:
What kind of exceptions are we expecting here?

(n.b. I thought Flask/sqlalchemy automatically rolled back when exceptions are thrown.)

Reply:
Thanks @legoktm, I suspect you are correct about it rolling back automatically. I was using the same logic/workflow that I found [within manage.py itself.](https://github.com/freedomofpress/securedrop/blob/develop/securedrop/manage.py#LL216C1-L219C34) Based on that, should this try/except block be removed within the branch? What about in manage.py? Or are they not roughly equivalent use cases?